### ISSUE TITLE
FunctionNaming false positive when the function's name equals to the return type's name with type arguments

### ISSUE THREAD

**Issue by arkivanov** [detekt/detekt#6601]
## Expected Behavior

Check passes.

## Observed Behavior

Error: `Function names should match the pattern: [a-z][a-zA-Z0-9]* [FunctionNaming]`.

## Steps to Reproduce

```kotlin
interface Foo<T>

fun <T> Foo(): Foo<T> = object : Foo<T> {}
```

## Your Environment
* Version of detekt used: 1.23.3
* Version of Gradle used (if applicable): 8.0.1
* Gradle scan link (add `--scan` option when running the gradle task): https://gradle.com/s/geo5n27nh25ne
* Operating System and version: macOS 13.5.2

**Comment by BraisGabin** [comment]
Why is this a false positive? The name of that function doesn't follow the pattern.

**Comment by t-kameyama** [comment]
> Why is this a false positive? The name of that function doesn't follow the pattern.

https://github.com/detekt/detekt/blob/2e703f04354c8e6b671fe7e67d443efa1882f643/detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/FunctionNamingSpec.kt#L68-L77

**Comment by arkivanov** [comment]
If I remove the `T` type parameter, then the code is not flagged.

```kotlin
interface Foo

fun Foo(): Foo = object : Foo {}
```

Also as per [Kotlin Coding conventions](https://kotlinlang.org/docs/coding-conventions.html#function-names):

> Exception: factory functions used to create instances of classes can have the same name as the abstract return type

From the coding convention:

```kotlin
interface Foo { /*...*/ }

class FooImpl : Foo { /*...*/ }

fun Foo(): Foo { return FooImpl() }
```
